### PR TITLE
refactor: use key/value cursor for serialization, retire ring buffer

### LINKED ISSUE #226: bug: abnormally high latency with gp3
configuration: ec2 c6id.2xlarge with EBS gp3 6000 IOPS 300MB/s 100GB

image

foyer lat

```plain
[3037s/7200s]
disk total iops: 3022.8
disk total throughput: 274.5 MiB/s
disk read iops: 969.0
disk read throughput: 45.5 MiB/s
disk write iops: 2053.8
disk write throughput: 229.0 MiB/s
insert iops: 2495.6/s
insert throughput: 224.1 MiB/s
insert lat p50: 78us
insert lat p90: 1564671us
insert lat p99: 1736703us
insert lat p999: 3506175us
insert lat p9999: 3883007us
insert lat p99999: 3997695us
insert lat pmax: 7405567us
get iops: 614.7/s
get miss: 5.28%
get throughput: 51.6 MiB/s
get hit lat p50: 2175us
get hit lat p90: 10879us
get hit lat p99: 14591us
get hit lat p999: 27775us
get hit lat p9999: 62719us
get hit lat p99999: 143359us
get hit lat pmax: 278527us
get miss lat p50: 54us
get miss lat p90: 435us
get miss lat p99: 2735us
get miss lat p999: 4767us
get miss lat p9999: 16319us
get miss lat p99999: 26879us
get miss lat pmax: 30591us
```

```bash
foyer-storage-bench --dir /ebs/foyer --capacity 81920 --region-size 64 --lookup-range 100000 --flushers 8 --reclaimers 4 --time 7200 --writers 4096 --w-rate 0.1 --ticket-insert-rate-limit 240 --readers 512 --r-rate 0.1 --runtime --entry-size-min 32768 --entry-size-max 153600 --metrics --io-size 65536 --ring-buffer-capacity 1500
```

### PR DESCRIPTION
**Pull request by MrCroxx**
## What's changed and what's your intention?

> Please explain **IN DETAIL** what the changes are in this PR and why they are needed. :D

As titled.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have passed `make check` and `make test` or `make all` in my local envirorment.

## Related issues or PRs (optional)
#216 
close #226

### CODE REVIEW

**Comment by MrCroxx** [comment]
# Benchmark

PR:

```
Total:
disk total iops: 16239.5
disk total throughput: 614.8 MiB/s
disk read iops: 8700.9
disk read throughput: 115.6 MiB/s
disk write iops: 7538.5
disk write throughput: 499.2 MiB/s
insert iops: 7514.1/s
insert throughput: 469.6 MiB/s
insert lat p50: 58us
insert lat p90: 149us
insert lat p99: 1447us
insert lat p999: 7391us
insert lat p9999: 19327us
insert lat p99999: 35583us
insert lat pmax: 40191us
get iops: 289454.7/s
get miss: 54.77%
get throughput: 8.0 GiB/s
get hit lat p50: 1247us
get hit lat p90: 2543us
get hit lat p99: 7231us
get hit lat p999: 32511us
get hit lat p9999: 44799us
get hit lat p99999: 52991us
get hit lat pmax: 68095us
get miss lat p50: 179us
get miss lat p90: 679us
get miss lat p99: 3215us
get miss lat p999: 7615us
get miss lat p9999: 14207us
get miss lat p99999: 25855us
get miss lat pmax: 40959us
```

main

```
Total:
disk total iops: 21664.7
disk total throughput: 685.2 MiB/s
disk read iops: 14151.4
disk read throughput: 187.9 MiB/s
disk write iops: 7513.2
disk write throughput: 497.3 MiB/s
insert iops: 7484.6/s
insert throughput: 467.8 MiB/s
insert lat p50: 83us
insert lat p90: 547us
insert lat p99: 5439us
insert lat p999: 20607us
insert lat p9999: 59391us
insert lat p99999: 73215us
insert lat pmax: 80383us
get iops: 288570.4/s
get miss: 54.65%
get throughput: 8.0 GiB/s
get hit lat p50: 1807us
get hit lat p90: 6143us
get hit lat p99: 25343us
get hit lat p999: 47359us
get hit lat p9999: 62975us
get hit lat p99999: 114687us
get hit lat pmax: 196607us
get miss lat p50: 313us
get miss lat p90: 1935us
get miss lat p99: 5919us
get miss lat p999: 13695us
get miss lat p9999: 29951us
get miss lat p99999: 43519us
get miss lat pmax: 65023us
```

No regression.

**Comment by MrCroxx** [comment]
@hzxa21 @Li0k

**Comment by Li0k** [comment]
I will review ASAP

File: foyer-common/src/code.rs [hunk 1]
```diff
@@ -433,7 +395,7 @@ impl std::io::Read for ArcVecU8Cursor {
         let slice = self.inner.as_ref().as_slice();
         let len = std::cmp::min(slice.len() - self.pos, buf.len());
         buf.put_slice(&slice[self.pos..self.pos + len]);
-        self.pos -= len;
+        self.pos += len;
```
Comment (Li0k) [inline foyer-common/src/code.rs:398]: Is it a bug introduced from previous PR ?

**Comment by Li0k** [comment]
coud you add some background info about the ring ?

File: foyer-common/src/code.rs (hunk 1 above)
Comment (MrCroxx) [reply foyer-common/src/code.rs:398]: Yep, this PR is part of a series to make each PR small. The previous one is not fully tested.

**Comment by MrCroxx** [comment]
> Could you add some background info about this PR?

The related issues informed some.

**Pull request merged**